crates/astria-conductor/src/executor/queue.rs

```diff
@@ -0,0 +1,504 @@
+
+        // check if the block is already present in the pending blocks
+        if let Some(pending_blocks) = self.pending_blocks.get(&height) {
+            if let Some(_block) = pending_blocks.get(&block_hash) {
+                return true;
+            }
+        }
+        // check if the block is already present in the soft blocks
+        if let Some(soft_block) = self.soft_blocks.get(&height) {
+            if soft_block.block_hash() == block_hash {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    fn is_block_a_parent(&mut self, block: SequencerBlockSubset) -> bool {
+        let block_hash = block.block_hash();
+        if let Some(child_blocks) = self.pending_blocks.get(&block.child_height()) {
+            let blocks = child_blocks.values();
+            for block in blocks {
+                if let Some(hash) = block.parent_hash() {
+                    if hash == block_hash {
+                        return true;
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    fn insert_to_pending_blocks(&mut self, block: SequencerBlockSubset) {
+        let height = block.height();
+        let block_hash = block.block_hash();
+
+        if let Some(pending_blocks) = self.pending_blocks.get_mut(&height) {
+            let mut new_map = pending_blocks.clone();
+            new_map.insert(block_hash, block);
+            self.pending_blocks.insert(height, new_map);
+        } else {
+            let mut new_map = HashMap::new();
+            new_map.insert(block_hash, block);
+            self.pending_blocks.insert(height, new_map);
+        }
+    }
+
+    // remove all data in the queue below a given height. this does not remove
+    // data from the soft queue, only the pending queue, and updates the head height.
+    // TODO: add error handling
+    fn remove_data_blow_height(&mut self, height: Height) {
+        // remove all data below the new incoming block from the pending data
+        let tmp_pending = self.pending_blocks.clone();
+        let mut pending_keys: Vec<&Height> = tmp_pending.keys().collect();
+        pending_keys.sort();
+        for key in pending_keys {
+            if *key < height {
+                self.pending_blocks.remove(key);
+            }
+        }
+    }
+
+    /// This function organizes the internal state of the queue based on the
+    /// tendermint/CometBTF fork choice rules.
+    ///
+    /// Once a block is added to the pending_blocks in the queue, this function
+    /// is called. It walks the panding blocks from lowest to highest height,
+    /// checking to see if there is a continues chain of blocks. For every block
+    /// that is a decendant of the most recent "soft" block, and has a direct
+    /// decendant, that block gets added to the `soft_blocks` BTreeMap and the
+    /// head height is updated.
+    /// after, and including, the most recent "soft" block that has a direct child
```

Comment:
Is this supposed to be somewhere in the above paragraph?

Reply:
this was a typo. it's been removed